packages/gatsby-link/src/index.js

```diff
@@ -131,59 +165,67 @@ class GatsbyLink extends React.Component {
       /* eslint-enable no-unused-vars */
       ...rest
     } = this.props
-
-    const LOCAL_URL = /^\/(?!\/)/
-    if (process.env.NODE_ENV !== `production` && !LOCAL_URL.test(to)) {
+    if (process.env.NODE_ENV !== `production` && !isLocalLink(to)) {
       console.warn(
         `External link ${to} was detected in a Link component. Use the Link component only for internal links. See: https://gatsby.dev/internal-links`
       )
     }
 
-    const prefixedTo = withPrefix(to)
-
     return (
-      <Link
-        to={prefixedTo}
-        state={state}
-        getProps={getProps}
-        innerRef={this.handleRef}
-        onMouseEnter={e => {
-          if (onMouseEnter) {
-            onMouseEnter(e)
-          }
-          ___loader.hovering(parsePath(to).pathname)
-        }}
-        onClick={e => {
-          if (onClick) {
-            onClick(e)
-          }
-
-          if (
-            e.button === 0 && // ignore right clicks
-            !this.props.target && // let browser handle "target=_blank"
-            !e.defaultPrevented && // onClick prevented default
-            !e.metaKey && // ignore clicks with modifier keys...
-            !e.altKey &&
-            !e.ctrlKey &&
-            !e.shiftKey
-          ) {
-            e.preventDefault()
-
-            let shouldReplace = replace
-            const isCurrent = encodeURI(to) === window.location.pathname
-            if (typeof replace !== `boolean` && isCurrent) {
-              shouldReplace = true
-            }
-
-            // Make sure the necessary scripts and data are
-            // loaded before continuing.
-            navigate(to, { state, replace: shouldReplace })
-          }
-
-          return true
+      <Location>
+        {({ location }) => {
+          const prefixedTo = rewriteLinkPath(to, location.pathname)
+          return isLocalLink(prefixedTo) ? (
```

Comment:
@ascorbic did this also add support for external links?

I'm asking because the [docs](https://www.gatsbyjs.org/docs/gatsby-link/#replace-a-tags-with-the-link-tag-for-local-links) have long warned against using <Link /> for external links.

I think it's a nice feature, until now I've had to implement a similar workaround in a wrapper around <Link /> to render either a Gatsby Link for local links or an anchor for others.

However this PR misses two things if this really is adding support for external links:

- we need to update the docs for it
- we need to remove this warning:

https://github.com/gatsbyjs/gatsby/blob/be86318a61b7ae21e1dda743d918700fee10e1cd/packages/gatsby-link/src/index.js#L169-L173

Reply:
@robinmetral This change handles them more gracefully, but we still discourage their use, which is why we still have the warning. It means we can change the behaviour in future without making it a breaking change.

Reply:
Thanks @ascorbic, got it. So as far as I understood, I can basically use Link for external links now, but I'll continue getting warnings.

What would be missing to make this a more officially supported feature? I can help if necessary, I'd love to get rid of my custom wrapper 😛

Reply:
You can continue to use them, but risk having us change behaviour in the future. For example, adding relative link support broke some existing links without leading slashes, which had worked but with warnings.